src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,190 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Agent.Sdk;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.BlobStore.Common;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent;
+using Microsoft.VisualStudio.Services.Content.Common;
+
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    internal class FileShareProvider: AgentService, IArtifactProvider
+    {
+        private readonly AgentTaskPluginExecutionContext context;
+        private readonly CallbackAppTraceSource tracer;
+        private const int defaultParallelCount = 1;
+        private string hostType;
+        // Default stream buffer size set in the existing file share implementation https://github.com/microsoft/azure-pipelines-agent/blob/ffb3a9b3e2eb5a1f34a0f45d0f2b8639740d37d3/src/Agent.Worker/Release/Artifacts/FileShareArtifact.cs#L154
+        private const int DefaultStreamBufferSize = 8192;
+
+        public FileShareProvider(AgentTaskPluginExecutionContext context, CallbackAppTraceSource tracer)
+        {
+            this.context = context;
+            this.tracer = tracer;
+            this.hostType = context.Variables.GetValueOrDefault("system.hosttype")?.Value;
+        }
+
+        public async Task DownloadSingleArtifactAsync(PipelineArtifactDownloadParameters downloadParameters, BuildArtifact buildArtifact, CancellationToken cancellationToken)
+        {
+            var downloadRootPath = Path.Combine(buildArtifact.Resource.Data, buildArtifact.Name);
```

Comment:
>buildArtifact.Name [](start = 77, length = 18)

is this tested for invalid chars?

Reply:
or malicious ones anyway

---
In reply to: [319237579](https://github.com/microsoft/azure-pipelines-agent/pull/2406#discussion_r319237579) [](ancestors = 319237579)

Reply:
Yes, it's being handled at the beginning of the code flow just like the rest of the pipeline artifact feature.